File: machine/cpp/machine/devices/gx/gpu_command_timing.cpp
#include "machine/devices/gx/gpu_command_timing.h"

namespace bmsx {
namespace {

i64 lineTicks(i32 x0, i32 y0, i32 x1, i32 y1, u32 drawingAreaTopLeftWord, u32 drawingAreaBottomRightWord, u8 interlacedRenderWord) {
	const i32 sourceLeft = x0 < x1 ? x0 : x1;
	const i32 sourceRight = x0 > x1 ? x0 : x1;
	const i32 sourceTop = y0 < y1 ? y0 : y1;
	const i32 sourceBottom = y0 > y1 ? y0 : y1;
	if (sourceRight - sourceLeft + 1 > 1024 || sourceBottom - sourceTop + 1 > 512) return 0;
	const i32 areaLeft = static_cast<i32>(gxGpuDrawingAreaLeft(drawingAreaTopLeftWord, drawingAreaBottomRightWord));
	const i32 areaTop = static_cast<i32>(gxGpuDrawingAreaTop(drawingAreaTopLeftWord, drawingAreaBottomRightWord));
	const i32 areaRight = static_cast<i32>(gxGpuDrawingAreaRightExclusive(drawingAreaTopLeftWord, drawingAreaBottomRightWord));
	const i32 areaBottom = static_cast<i32>(gxGpuDrawingAreaBottomExclusive(drawingAreaTopLeftWord, drawingAreaBottomRightWord));
	const i32 left = sourceLeft > areaLeft ? sourceLeft : areaLeft;
	const i32 top = sourceTop > areaTop ? sourceTop : areaTop;
	const i32 right = sourceRight + 1 < areaRight ? sourceRight + 1 : areaRight;
	const i32 bottom = sourceBottom + 1 < areaBottom ? sourceBottom + 1 : areaBottom;
	if (left >= right || top >= bottom) return 0;
	const i32 width = right - left;
	i32 height = bottom - top;
	if ((interlacedRenderWord & GX_GPU_INTERLACED_RENDER_ENABLE) != 0u) {
		height >>= 1;
		if (height == 0) height = 1;
	}
	return width > height ? width : height;
}

i64 triangleTicks(
	i32 x0,
	i32 y0,
	i32 x1,
	i32 y1,
	i32 x2,
	i32 y2,
	bool textured,
	bool semiTransparent,
	u32 drawingAreaTopLeftWord,
	u32 drawingAreaBottomRightWord,
	u32 maskBitModeWord,
	u8 interlacedRenderWord) {
	const i32 min12X = x1 < x2 ? x1 : x2;
	const i32 max12X = x1 > x2 ? x1 : x2;
	const i32 min12Y = y1 < y2 ? y1 : y2;
	const i32 max12Y = y1 > y2 ? y1 : y2;
	const i32 sourceLeft = x0 < min12X ? x0 : min12X;
	const i32 sourceRight = x0 > max12X ? x0 : max12X;
	const i32 sourceTop = y0 < min12Y ? y0 : min12Y;
	const i32 sourceBottom = y0 > max12Y ? y0 : max12Y;
	if (sourceRight - sourceLeft + 1 > 1024 || sourceBottom - sourceTop + 1 > 512) return 0;
	const i32 left = static_cast<i32>(gxGpuDrawingAreaLeft(drawingAreaTopLeftWord, drawingAreaBottomRightWord));
	const i32 top = static_cast<i32>(gxGpuDrawingAreaTop(drawingAreaTopLeftWord, drawingAreaBottomRightWord));
	const i32 right = static_cast<i32>(gxGpuDrawingAreaRightExclusive(drawingAreaTopLeftWord, drawingAreaBottomRightWord));
	const i32 bottom = static_cast<i32>(gxGpuDrawingAreaBottomExclusive(drawingAreaTopLeftWord, drawingAreaBottomRightWord));
	if (left >= right || top >= bottom) return 0;
	const i32 rightEdge = right - 1;
	const i32 bottomEdge = bottom - 1;
	x0 = x0 < left ? left : (x0 > rightEdge ? rightEdge : x0);
	y0 = y0 < top ? top : (y0 > bottomEdge ? bottomEdge : y0);
	x1 = x1 < left ? left : (x1 > rightEdge ? rightEdge : x1);
	y1 = y1 < top ? top : (y1 > bottomEdge ? bottomEdge : y1);
	x2 = x2 < left ? left : (x2 > rightEdge ? rightEdge : x2);
	y2 = y2 < top ? top : (y2 > bottomEdge ? bottomEdge : y2);
	i64 doubleArea = static_cast<i64>(x0) * y1 + static_cast<i64>(x1) * y2 + static_cast<i64>(x2) * y0
		- static_cast<i64>(x0) * y2 - static_cast<i64>(x1) * y0 - static_cast<i64>(x2) * y1;
	if (doubleArea < 0) doubleArea = -doubleArea;
	i64 pixels = doubleArea >> 1u;
	if (textured) pixels += pixels;
	if (semiTransparent || (maskBitModeWord & 0x02u) != 0u) pixels += (pixels + 1) >> 1u;
	if ((interlacedRenderWord & GX_GPU_INTERLACED_RENDER_ENABLE) != 0u) pixels >>= 1u;
	return pixels;
}

i64 polygonTicks(
	u8 opcode,
	const u32* words,
	size_t wordStart,
	u32 drawModeWord,
	u32 drawingAreaTopLeftWord,
	u32 drawingAreaBottomRightWord,
	u32 drawingOffsetWord,
	u32 maskBitModeWord,
	u8 interlacedRenderWord) {
	const bool packetTextured = (opcode & 0x04u) != 0u;
	const bool textured = packetTextured && (drawModeWord & GX_GPU_DRAW_MODE_TEXTURE_DISABLE) == 0u;
	const bool quad = (opcode & 0x08u) != 0u;
	const bool gouraud = (opcode & 0x10u) != 0u;
	i64 ticks = (quad ? 82 : 46) + (gouraud ? 288 : 0) + (textured ? (gouraud ? 162 : 180) : 0);
	const size_t stride = 1u + (packetTextured ? 1u : 0u) + (gouraud ? 1u : 0u);
	const i32 offsetX = gxGpuSigned11(drawingOffsetWord);
	const i32 offsetY = gxGpuSigned11(drawingOffsetWord >> 11u);
	const u32 vertex0 = words[wordStart + 1u];
	const u32 vertex1 = words[wordStart + 1u + stride];
	const u32 vertex2 = words[wordStart + 1u + stride * 2u];
	const i32 x0 = gxGpuSigned11(vertex0) + offsetX;
	const i32 y0 = gxGpuSigned11(vertex0 >> 16u) + offsetY;
	const i32 x1 = gxGpuSigned11(vertex1) + offsetX;
	const i32 y1 = gxGpuSigned11(vertex1 >> 16u) + offsetY;
	const i32 x2 = gxGpuSigned11(vertex2) + offsetX;
	const i32 y2 = gxGpuSigned11(vertex2 >> 16u) + offsetY;
	const bool semiTransparent = (opcode & 0x02u) != 0u;
	ticks += triangleTicks(x0, y0, x1, y1, x2, y2, textured, semiTransparent, drawingAreaTopLeftWord, drawingAreaBottomRightWord, maskBitModeWord, interlacedRenderWord);
	if (quad) {
		const u32 vertex3 = words[wordStart + 1u + stride * 3u];
		const i32 x3 = gxGpuSigned11(vertex3) + offsetX;
		const i32 y3 = gxGpuSigned11(vertex3 >> 16u) + offsetY;
		ticks += triangleTicks(x2, y2, x1, y1, x3, y3, textured, semiTransparent, drawingAreaTopLeftWord, drawingAreaBottomRightWord, maskBitModeWord, interlacedRenderWord);
	}
	return ticks;
}

i64 rectangleTicks(
	u8 opcode,
	const u32* words,
	size_t wordStart,
	u32 wordCount,
	u32 drawModeWord,
	u32 drawingAreaTopLeftWord,
	u32 drawingAreaBottomRightWord,
	u32 drawingOffsetWord,
	u32 maskBitModeWord,
	u8 interlacedRenderWord) {
	u32 width;
	u32 height;
	switch (opcode & 0x18u) {
		case 0x08u: width = 1u; height = 1u; break;
		case 0x10u: width = 8u; height = 8u; break;
		case 0x18u: width = 16u; height = 16u; break;
		default: {
			const u32 sizeWord = words[wordStart + wordCount - 1u];
			width = sizeWord & 0x3ffu;
			height = (sizeWord >> 16u) & 0x1ffu;
			break;
		}
	}
	const u32 positionWord = words[wordStart + 1u];
	const i32 x = gxGpuSigned11(positionWord) + gxGpuSigned11(drawingOffsetWord);
	const i32 y = gxGpuSigned11(positionWord >> 16u) + gxGpuSigned11(drawingOffsetWord >> 11u);
	const i32 areaLeft = static_cast<i32>(gxGpuDrawingAreaLeft(drawingAreaTopLeftWord, drawingAreaBottomRightWord));
	const i32 areaTop = static_cast<i32>(gxGpuDrawingAreaTop(drawingAreaTopLeftWord, drawingAreaBottomRightWord));
	const i32 areaRight = static_cast<i32>(gxGpuDrawingAreaRightExclusive(drawingAreaTopLeftWord, drawingAreaBottomRightWord));
	const i32 areaBottom = static_cast<i32>(gxGpuDrawingAreaBottomExclusive(drawingAreaTopLeftWord, drawingAreaBottomRightWord));
	const i32 left = x > areaLeft ? x : areaLeft;
	const i32 top = y > areaTop ? y : areaTop;
	const i32 right = x + static_cast<i32>(width) < areaRight ? x + static_cast<i32>(width) : areaRight;
	const i32 bottom = y + static_cast<i32>(height) < areaBottom ? y + static_cast<i32>(height) : areaBottom;
	if (left >= right || top >= bottom) return 16;
	const u32 drawnWidth = static_cast<u32>(right - left);
	u32 drawnHeight = static_cast<u32>(bottom - top);
	u32 ticksPerRow = drawnWidth;
	const bool textured = (opcode & 0x04u) != 0u && (drawModeWord & GX_GPU_DRAW_MODE_TEXTURE_DISABLE) == 0u;
	if (textured) {
		switch ((drawModeWord >> 7u) & 0x3u) {
			case GX_GPU_TEXTURE_MODE_PALETTE4:
				ticksPerRow += drawnWidth;
				break;
			case GX_GPU_TEXTURE_MODE_PALETTE8:
				if (drawnWidth > 128u) {
					ticksPerRow += (drawnWidth >> 2u) * 8u;
				} else if (drawnWidth * drawnHeight > 2048u) {
					ticksPerRow += (drawnWidth >> 2u) * (4u * (128u / drawnWidth));
				} else {
					ticksPerRow += drawnWidth;
				}
				break;
			case GX_GPU_TEXTURE_MODE_DIRECT16:
			default:
				if (drawnWidth > 128u) {
					ticksPerRow += (drawnWidth >> 1u) * 8u;
				} else if (drawnWidth * drawnHeight > 1024u) {
					ticksPerRow += (drawnWidth >> 2u) * (8u * (128u / drawnWidth));
				} else {
					ticksPerRow += drawnWidth;
				}
				break;
		}
	}
	if ((opcode & 0x02u) != 0u || (maskBitModeWord & 0x02u) != 0u) ticksPerRow += (drawnWidth + 1u) >> 1u;
	if ((interlacedRenderWord & GX_GPU_INTERLACED_RENDER_ENABLE) != 0u) {
		drawnHeight >>= 1u;
		if (drawnHeight == 0u) drawnHeight = 1u;
	}
	return 16 + static_cast<i64>(ticksPerRow) * drawnHeight;
}

i64 polylineTicks(
	u8 opcode,
	const u32* words,
	size_t wordStart,
	u32 wordCount,
	u32 drawingAreaTopLeftWord,
	u32 drawingAreaBottomRightWord,
	u32 drawingOffsetWord,
	u8 interlacedRenderWord) {
	const size_t stride = (opcode & 0x10u) != 0u ? 2u : 1u;
	const i32 offsetX = gxGpuSigned11(drawingOffsetWord);
	const i32 offsetY = gxGpuSigned11(drawingOffsetWord >> 11u);
	size_t positionIndex = wordStart + 1u;
	u32 positionWord = words[positionIndex];
	i32 x0 = gxGpuSigned11(positionWord) + offsetX;
	i32 y0 = gxGpuSigned11(positionWord >> 16u) + offsetY;
	i64 ticks = 16;
	positionIndex += stride;
	while (positionIndex < wordStart + wordCount) {
		positionWord = words[positionIndex];
		const i32 x1 = gxGpuSigned11(positionWord) + offsetX;
		const i32 y1 = gxGpuSigned11(positionWord >> 16u) + offsetY;
		ticks += lineTicks(x0, y0, x1, y1, drawingAreaTopLeftWord, drawingAreaBottomRightWord, interlacedRenderWord);
		x0 = x1;
		y0 = y1;
		positionIndex += stride;
	}
	return ticks;
}

} // namespace

i64 gxGpuCommandTicks(
	u8 kind,
	u8 opcode,
	const u32* words,
	size_t wordStart,
	u32 wordCount,
	u32 drawModeWord,
	u32 drawingAreaTopLeftWord,
	u32 drawingAreaBottomRightWord,
	u32 drawingOffsetWord,
	u32 maskBitModeWord,
	u8 interlacedRenderWord) {
	switch (kind) {
		case GX_GPU_COMMAND_DRAW_POLYGON:
			return polygonTicks(opcode, words, wordStart, drawModeWord, drawingAreaTopLeftWord, drawingAreaBottomRightWord, drawingOffsetWord, maskBitModeWord, interlacedRenderWord);
		case GX_GPU_COMMAND_DRAW_LINE: {
			const i32 offsetX = gxGpuSigned11(drawingOffsetWord);
			const i32 offsetY = gxGpuSigned11(drawingOffsetWord >> 11u);
			const u32 first = words[wordStart + 1u];
			const u32 second = words[wordStart + ((opcode & 0x10u) != 0u ? 3u : 2u)];
			return lineTicks(gxGpuSigned11(first) + offsetX, gxGpuSigned11(first >> 16u) + offsetY, gxGpuSigned11(second) + offsetX, gxGpuSigned11(second >> 16u) + offsetY, drawingAreaTopLeftWord, drawingAreaBottomRightWord, interlacedRenderWord);
		}
		case GX_GPU_COMMAND_DRAW_POLYLINE:
			return polylineTicks(opcode, words, wordStart, wordCount, drawingAreaTopLeftWord, drawingAreaBottomRightWord, drawingOffsetWord, interlacedRenderWord);
		case GX_GPU_COMMAND_DRAW_RECTANGLE:
			return rectangleTicks(opcode, words, wordStart, wordCount, drawModeWord, drawingAreaTopLeftWord, drawingAreaBottomRightWord, drawingOffsetWord, maskBitModeWord, interlacedRenderWord);
		case GX_GPU_COMMAND_FILL_RECTANGLE: {
			const u32 sizeWord = words[wordStart + 2u];
			const u32 width = ((sizeWord & 0x3ffu) + 0x0fu) & ~0x0fu;
			const u32 height = (sizeWord >> 16u) & 0x1ffu;
			return 46 + static_cast<i64>((width >> 3u) + 9u) * height;
		}
		case GX_GPU_COMMAND_COPY_VRAM_TO_VRAM: {
			const u32 sizeWord = words[wordStart + 3u];
			const u32 width = (((sizeWord & 0xffffu) - 1u) & 0x3ffu) + 1u;
			const u32 height = ((((sizeWord >> 16u) & 0xffffu) - 1u) & 0x1ffu) + 1u;
			return static_cast<i64>(width) * height * 2;
		}
		case GX_GPU_COMMAND_UPLOAD_CPU_TO_VRAM:
		case GX_GPU_COMMAND_READ_VRAM_TO_CPU:
		default:
			return 1;
	}
}

} // namespace bmsx

File: machine/ts/machine/devices/gx/gpu_command_timing.ts
import {
	GX_GPU_COMMAND_COPY_VRAM_TO_VRAM,
	GX_GPU_COMMAND_DRAW_LINE,
	GX_GPU_COMMAND_DRAW_POLYGON,
	GX_GPU_COMMAND_DRAW_POLYLINE,
	GX_GPU_COMMAND_DRAW_RECTANGLE,
	GX_GPU_COMMAND_FILL_RECTANGLE,
	GX_GPU_COMMAND_READ_VRAM_TO_CPU,
	GX_GPU_COMMAND_UPLOAD_CPU_TO_VRAM,
	GX_GPU_DRAW_MODE_TEXTURE_DISABLE,
	GX_GPU_INTERLACED_RENDER_ENABLE,
	GX_GPU_TEXTURE_MODE_DIRECT16,
	GX_GPU_TEXTURE_MODE_PALETTE4,
	GX_GPU_TEXTURE_MODE_PALETTE8,
	gxGpuDrawingAreaBottomExclusive,
	gxGpuDrawingAreaLeft,
	gxGpuDrawingAreaRightExclusive,
	gxGpuDrawingAreaTop,
	gxGpuSigned11,
} from './gpu_command_buffer';

export const GX_GPU_COMMAND_TICKS_PER_CPU_CYCLE = 2;

function lineTicks(
	x0: number,
	y0: number,
	x1: number,
	y1: number,
	drawingAreaTopLeftWord: number,
	drawingAreaBottomRightWord: number,
	interlacedRenderWord: number,
): number {
	const sourceLeft = x0 < x1 ? x0 : x1;
	const sourceRight = x0 > x1 ? x0 : x1;
	const sourceTop = y0 < y1 ? y0 : y1;
	const sourceBottom = y0 > y1 ? y0 : y1;
	if (sourceRight - sourceLeft + 1 > 1024 || sourceBottom - sourceTop + 1 > 512) return 0;
	const areaLeft = gxGpuDrawingAreaLeft(drawingAreaTopLeftWord, drawingAreaBottomRightWord);
	const areaTop = gxGpuDrawingAreaTop(drawingAreaTopLeftWord, drawingAreaBottomRightWord);
	const areaRight = gxGpuDrawingAreaRightExclusive(drawingAreaTopLeftWord, drawingAreaBottomRightWord);
	const areaBottom = gxGpuDrawingAreaBottomExclusive(drawingAreaTopLeftWord, drawingAreaBottomRightWord);
	const left = sourceLeft > areaLeft ? sourceLeft : areaLeft;
	const top = sourceTop > areaTop ? sourceTop : areaTop;
	const right = sourceRight + 1 < areaRight ? sourceRight + 1 : areaRight;
	const bottom = sourceBottom + 1 < areaBottom ? sourceBottom + 1 : areaBottom;
	if (left >= right || top >= bottom) return 0;
	const width = right - left;
	let height = bottom - top;
	if ((interlacedRenderWord & GX_GPU_INTERLACED_RENDER_ENABLE) !== 0) {
		height >>= 1;
		if (height === 0) height = 1;
	}
	return width > height ? width : height;
}

function triangleTicks(
	x0: number,
	y0: number,
	x1: number,
	y1: number,
	x2: number,
	y2: number,
	textured: boolean,
	semiTransparent: boolean,
	drawingAreaTopLeftWord: number,
	drawingAreaBottomRightWord: number,
	maskBitModeWord: number,
	interlacedRenderWord: number,
): number {
	const min12X = x1 < x2 ? x1 : x2;
	const max12X = x1 > x2 ? x1 : x2;
	const min12Y = y1 < y2 ? y1 : y2;
	const max12Y = y1 > y2 ? y1 : y2;
	const sourceLeft = x0 < min12X ? x0 : min12X;
	const sourceRight = x0 > max12X ? x0 : max12X;
	const sourceTop = y0 < min12Y ? y0 : min12Y;
	const sourceBottom = y0 > max12Y ? y0 : max12Y;
	if (sourceRight - sourceLeft + 1 > 1024 || sourceBottom - sourceTop + 1 > 512) return 0;
	const left = gxGpuDrawingAreaLeft(drawingAreaTopLeftWord, drawingAreaBottomRightWord);
	const top = gxGpuDrawingAreaTop(drawingAreaTopLeftWord, drawingAreaBottomRightWord);
	const right = gxGpuDrawingAreaRightExclusive(drawingAreaTopLeftWord, drawingAreaBottomRightWord);
	const bottom = gxGpuDrawingAreaBottomExclusive(drawingAreaTopLeftWord, drawingAreaBottomRightWord);
	if (left >= right || top >= bottom) return 0;
	const rightEdge = right - 1;
	const bottomEdge = bottom - 1;
	x0 = x0 < left ? left : (x0 > rightEdge ? rightEdge : x0);
	y0 = y0 < top ? top : (y0 > bottomEdge ? bottomEdge : y0);
	x1 = x1 < left ? left : (x1 > rightEdge ? rightEdge : x1);
	y1 = y1 < top ? top : (y1 > bottomEdge ? bottomEdge : y1);
	x2 = x2 < left ? left : (x2 > rightEdge ? rightEdge : x2);
	y2 = y2 < top ? top : (y2 > bottomEdge ? bottomEdge : y2);
	let doubleArea = x0 * y1 + x1 * y2 + x2 * y0 - x0 * y2 - x1 * y0 - x2 * y1;
	if (doubleArea < 0) doubleArea = -doubleArea;
	let pixels = (doubleArea - (doubleArea & 1)) / 2;
	if (textured) pixels += pixels;
	if (semiTransparent || (maskBitModeWord & 0x02) !== 0) pixels += (pixels + 1) >> 1;
	if ((interlacedRenderWord & GX_GPU_INTERLACED_RENDER_ENABLE) !== 0) pixels >>= 1;
	return pixels;
}

function polygonTicks(
	opcode: number,
	words: ArrayLike<number>,
	wordStart: number,
	drawModeWord: number,
	drawingAreaTopLeftWord: number,
	drawingAreaBottomRightWord: number,
	drawingOffsetWord: number,
	maskBitModeWord: number,
	interlacedRenderWord: number,
): number {
	const packetTextured = (opcode & 0x04) !== 0;
	const textured = packetTextured && (drawModeWord & GX_GPU_DRAW_MODE_TEXTURE_DISABLE) === 0;
	const quad = (opcode & 0x08) !== 0;
	const gouraud = (opcode & 0x10) !== 0;
	let ticks = (quad ? 82 : 46) + (gouraud ? 288 : 0) + (textured ? (gouraud ? 162 : 180) : 0);
	const stride = 1 + (packetTextured ? 1 : 0) + (gouraud ? 1 : 0);
	const offsetX = gxGpuSigned11(drawingOffsetWord);
	const offsetY = gxGpuSigned11(drawingOffsetWord >>> 11);
	const vertex0 = words[wordStart + 1]!;
	const vertex1 = words[wordStart + 1 + stride]!;
	const vertex2 = words[wordStart + 1 + stride * 2]!;
	const x0 = gxGpuSigned11(vertex0) + offsetX;
	const y0 = gxGpuSigned11(vertex0 >>> 16) + offsetY;
	const x1 = gxGpuSigned11(vertex1) + offsetX;
	const y1 = gxGpuSigned11(vertex1 >>> 16) + offsetY;
	const x2 = gxGpuSigned11(vertex2) + offsetX;
	const y2 = gxGpuSigned11(vertex2 >>> 16) + offsetY;
	const semiTransparent = (opcode & 0x02) !== 0;
	ticks += triangleTicks(x0, y0, x1, y1, x2, y2, textured, semiTransparent, drawingAreaTopLeftWord, drawingAreaBottomRightWord, maskBitModeWord, interlacedRenderWord);
	if (quad) {
		const vertex3 = words[wordStart + 1 + stride * 3]!;
		const x3 = gxGpuSigned11(vertex3) + offsetX;
		const y3 = gxGpuSigned11(vertex3 >>> 16) + offsetY;
		ticks += triangleTicks(x2, y2, x1, y1, x3, y3, textured, semiTransparent, drawingAreaTopLeftWord, drawingAreaBottomRightWord, maskBitModeWord, interlacedRenderWord);
	}
	return ticks;
}

function rectangleTicks(
	opcode: number,
	words: ArrayLike<number>,
	wordStart: number,
	wordCount: number,
	drawModeWord: number,
	drawingAreaTopLeftWord: number,
	drawingAreaBottomRightWord: number,
	drawingOffsetWord: number,
	maskBitModeWord: number,
	interlacedRenderWord: number,
): number {
	let width: number;
	let height: number;
	switch (opcode & 0x18) {
		case 0x08:
			width = 1;
			height = 1;
			break;
		case 0x10:
			width = 8;
			height = 8;
			break;
		case 0x18:
			width = 16;
			height = 16;
			break;
		default: {
			const sizeWord = words[wordStart + wordCount - 1]!;
			width = sizeWord & 0x3ff;
			height = (sizeWord >>> 16) & 0x1ff;
			break;
		}
	}
	const positionWord = words[wordStart + 1]!;
	const x = gxGpuSigned11(positionWord) + gxGpuSigned11(drawingOffsetWord);
	const y = gxGpuSigned11(positionWord >>> 16) + gxGpuSigned11(drawingOffsetWord >>> 11);
	const areaLeft = gxGpuDrawingAreaLeft(drawingAreaTopLeftWord, drawingAreaBottomRightWord);
	const areaTop = gxGpuDrawingAreaTop(drawingAreaTopLeftWord, drawingAreaBottomRightWord);
	const areaRight = gxGpuDrawingAreaRightExclusive(drawingAreaTopLeftWord, drawingAreaBottomRightWord);
	const areaBottom = gxGpuDrawingAreaBottomExclusive(drawingAreaTopLeftWord, drawingAreaBottomRightWord);
	const left = x > areaLeft ? x : areaLeft;
	const top = y > areaTop ? y : areaTop;
	const right = x + width < areaRight ? x + width : areaRight;
	const bottom = y + height < areaBottom ? y + height : areaBottom;
	if (left >= right || top >= bottom) return 16;
	const drawnWidth = right - left;
	let drawnHeight = bottom - top;
	let ticksPerRow = drawnWidth;
	const textured = (opcode & 0x04) !== 0 && (drawModeWord & GX_GPU_DRAW_MODE_TEXTURE_DISABLE) === 0;
	if (textured) {
		switch ((drawModeWord >>> 7) & 0x3) {
			case GX_GPU_TEXTURE_MODE_PALETTE4:
				ticksPerRow += drawnWidth;
				break;
			case GX_GPU_TEXTURE_MODE_PALETTE8:
				if (drawnWidth > 128) {
					ticksPerRow += (drawnWidth >> 2) * 8;
				} else if (drawnWidth * drawnHeight > 2048) {
					const rowsPerCache = (128 - (128 % drawnWidth)) / drawnWidth;
					ticksPerRow += (drawnWidth >> 2) * (4 * rowsPerCache);
				} else {
					ticksPerRow += drawnWidth;
				}
				break;
			case GX_GPU_TEXTURE_MODE_DIRECT16:
			default:
				if (drawnWidth > 128) {
					ticksPerRow += (drawnWidth >> 1) * 8;
				} else if (drawnWidth * drawnHeight > 1024) {
					const rowsPerCache = (128 - (128 % drawnWidth)) / drawnWidth;
					ticksPerRow += (drawnWidth >> 2) * (8 * rowsPerCache);
				} else {
					ticksPerRow += drawnWidth;
				}
				break;
		}
	}
	if ((opcode & 0x02) !== 0 || (maskBitModeWord & 0x02) !== 0) ticksPerRow += (drawnWidth + 1) >> 1;
	if ((interlacedRenderWord & GX_GPU_INTERLACED_RENDER_ENABLE) !== 0) {
		drawnHeight >>= 1;
		if (drawnHeight === 0) drawnHeight = 1;
	}
	return 16 + ticksPerRow * drawnHeight;
}

function polylineTicks(
	opcode: number,
	words: ArrayLike<number>,
	wordStart: number,
	wordCount: number,
	drawingAreaTopLeftWord: number,
	drawingAreaBottomRightWord: number,
	drawingOffsetWord: number,
	interlacedRenderWord: number,
): number {
	const stride = (opcode & 0x10) !== 0 ? 2 : 1;
	const offsetX = gxGpuSigned11(drawingOffsetWord);
	const offsetY = gxGpuSigned11(drawingOffsetWord >>> 11);
	let positionIndex = wordStart + 1;
	let positionWord = words[positionIndex]!;
	let x0 = gxGpuSigned11(positionWord) + offsetX;
	let y0 = gxGpuSigned11(positionWord >>> 16) + offsetY;
	let ticks = 16;
	positionIndex += stride;
	while (positionIndex < wordStart + wordCount) {
		positionWord = words[positionIndex]!;
		const x1 = gxGpuSigned11(positionWord) + offsetX;
		const y1 = gxGpuSigned11(positionWord >>> 16) + offsetY;
		ticks += lineTicks(x0, y0, x1, y1, drawingAreaTopLeftWord, drawingAreaBottomRightWord, interlacedRenderWord);
		x0 = x1;
		y0 = y1;
		positionIndex += stride;
	}
	return ticks;
}

export function gxGpuCommandTicks(
	kind: number,
	opcode: number,
	words: ArrayLike<number>,
	wordStart: number,
	wordCount: number,
	drawModeWord: number,
	drawingAreaTopLeftWord: number,
	drawingAreaBottomRightWord: number,
	drawingOffsetWord: number,
	maskBitModeWord: number,
	interlacedRenderWord: number,
): number {
	switch (kind) {
		case GX_GPU_COMMAND_DRAW_POLYGON:
			return polygonTicks(opcode, words, wordStart, drawModeWord, drawingAreaTopLeftWord, drawingAreaBottomRightWord, drawingOffsetWord, maskBitModeWord, interlacedRenderWord);
		case GX_GPU_COMMAND_DRAW_LINE: {
			const offsetX = gxGpuSigned11(drawingOffsetWord);
			const offsetY = gxGpuSigned11(drawingOffsetWord >>> 11);
			const first = words[wordStart + 1]!;
			const second = words[wordStart + ((opcode & 0x10) !== 0 ? 3 : 2)]!;
			return lineTicks(gxGpuSigned11(first) + offsetX, gxGpuSigned11(first >>> 16) + offsetY, gxGpuSigned11(second) + offsetX, gxGpuSigned11(second >>> 16) + offsetY, drawingAreaTopLeftWord, drawingAreaBottomRightWord, interlacedRenderWord);
		}
		case GX_GPU_COMMAND_DRAW_POLYLINE:
			return polylineTicks(opcode, words, wordStart, wordCount, drawingAreaTopLeftWord, drawingAreaBottomRightWord, drawingOffsetWord, interlacedRenderWord);
		case GX_GPU_COMMAND_DRAW_RECTANGLE:
			return rectangleTicks(opcode, words, wordStart, wordCount, drawModeWord, drawingAreaTopLeftWord, drawingAreaBottomRightWord, drawingOffsetWord, maskBitModeWord, interlacedRenderWord);
		case GX_GPU_COMMAND_FILL_RECTANGLE: {
			const sizeWord = words[wordStart + 2]!;
			const width = ((sizeWord & 0x3ff) + 0x0f) & ~0x0f;
			const height = (sizeWord >>> 16) & 0x1ff;
			return 46 + ((width >> 3) + 9) * height;
		}
		case GX_GPU_COMMAND_COPY_VRAM_TO_VRAM: {
			const sizeWord = words[wordStart + 3]!;
			const width = (((sizeWord & 0xffff) - 1) & 0x3ff) + 1;
			const height = ((((sizeWord >>> 16) & 0xffff) - 1) & 0x1ff) + 1;
			return width * height * 2;
		}
		case GX_GPU_COMMAND_UPLOAD_CPU_TO_VRAM:
		case GX_GPU_COMMAND_READ_VRAM_TO_CPU:
			return 1;
		default:
			return 1;
	}
}
